pkg/subscriptions/subscriptions.go

```diff
@@ -306,43 +321,92 @@ func addSubscriptionsFromMountsFile(filePath, mountLabel, containerRunDir string
 	return mounts, nil
 }
 
-// addFIPSModeSubscription adds mounts to the `mounts` slice that are needed for the container to run openssl in FIPs mode
-// (i.e: be FIPs compliant).
-// It should only be called if /etc/system-fips exists on host.
-// It primarily does two things:
-//   - creates /run/secrets/system-fips in the container root filesystem, and adds it to the `mounts` slice.
-//   - If `/etc/crypto-policies/back-ends` already exists inside of the container, it creates
-//     `/usr/share/crypto-policies/back-ends/FIPS` inside the container as well.
-//     It is done from within the container to ensure to avoid policy incompatibility between the container and host.
-func addFIPSModeSubscription(mounts *[]rspec.Mount, containerRunDir, mountPoint, mountLabel string, uid, gid int) error {
+func containerHasEtcSystemFips(subscriptionsDir, mountPoint string) (bool, error) {
+	fipsFileHost := filepath.Join(mountPoint, "etc/system-fips")
+	if fileutils.Lexists(fipsFileHost) != nil {
```

Comment:
If the container's /etc is a symlink it can escape on the host, this must use the securejoin package as well. untested: but I think this can be chnaged to securejoin.OpenInRoot() and then call readlinkat() with an empty path on it, per man page it should then resolve the dirfd argument directly.

Reply:
ReadlinkAt would return the contents of the link, but not _always_ correctly resolve it to the destination path (e.g. if the symlink contents contained are something like `../etc/../`…, we still need a symlink resolution operation.

I do agree this also needs protection against container escape.

Reply:
The issue here is that `securejoin` follows the `etc/system-fips` symlink, but I want to test whether that symlink exists and not actually follow it. In fact, this code shouldn't fail if the symlink exists but points to a non-existent file. Let me check whether that can be done with `securejoin`.

The code below is already safe against container escapes when checking whether `$container/etc/system-fips` points to `$container/run/secrets/system-fips`. I just need to know whether `$container/etc/system-fips` exists because the error handling differs based on that.

Maybe the simplest solution here is to do
```go
containerEtc := securejoin.SecureJoin(mountPoint, "etc")
if fileutils.Lexists(filepath.Join(containerEtc, "system-fips")) {
```
That should be secure against a container escape using a symlink `/etc`, and because `Lexists()` doesn't follow symlinks, it should also be safe against container escapes with the `etc/system-fips` file.

Reply:
Yes, that should be secure (assuming the filesystem is not being concurrently modified, and that assumption should be satisfied here before starting the container).

Reply:
That should work but I still think that makes the process more complicated then just resolving the path directly,
checking if link to then follow the link always feels odd to me, just try to resolve and handle 
```
fd,err := securejoin.OpenInRoot(mountPoint,"etc/system-fips")
if err != nil {... handle ENOENT}

pathBuf := make([]byte, 4096)
n,err := unix.Readlinkat(fd,"",pathBuf)
if err != nil {
    // handle EINVAL -> no link
}
path = string(pathBuf[:n])
//check if path

```

If we are really concerned about ` ../etc/../…` links (I would argue they are not sane anyway and so I do not see why we should support them) I think we could just clean `path` with `filepath.Clean()`

Reply:
That still leaves two valid values for pathBuf, `/run/secrets/system-fips` and `../run/secrets/system-fips`. It feels a lot cleaner to me to resolve the symlink and then check whether it points to where we expect it to than just reading the symlink.

To put it a different way, would you expect a difference of behavior in podman when you change your container image's `/etc/system-fips` symlink from `/run/secrets/system-fips` to `../run/secrets/system-fips` or (for whatever reason) `../../../../../../run/secrets/system-fips`?

Reply:
I mean we could try to parse that to if needed. Anyway I am fine with the current version if you want that, though the CONFIG_CRYPTO_FIPS abive is not yet addressed

Reply:
While we would try to parse that, I don't think it would be simpler than the current code.

I'll address the CONFIG_CRYPTO_FIPS flag in a second.